ophyd/signal.py

```diff
@@ -1182,52 +1202,66 @@ def _pv_access_callback(self, read_access, write_access, pv):
             return
 
         md_update = {}
-        if pv.pvname == self._read_pvname:
+        if pv.pvname == self.pvname:
             md_update['read_access'] = read_access
 
-        if pv.pvname is self.setpoint_pvname:
+        if pv.pvname == self.setpoint_pvname:
             md_update['write_access'] = write_access
 
         if md_update:
             self._metadata.update(**md_update)
 
         if self.connected:
-            self._run_subs(sub_type=self.SUB_META, **self._metadata)
+            self._run_metadata_callbacks()
 
         super()._pv_access_callback(read_access, write_access, pv)
         self._set_event_if_ready()
 
-    def _metadata_changed(self, pvname, cl_metadata, *, require_timestamp=False,
-                          update=True):
+    def _metadata_changed(self, pvname, cl_metadata, *, from_monitor,
+                          require_timestamp=False, update=True):
         'Metadata for one PV has changed'
         metadata = super()._metadata_changed(
-            pvname, cl_metadata, update=update,
+            pvname, cl_metadata, from_monitor=from_monitor, update=update,
             require_timestamp=require_timestamp)
 
-        if (self.setpoint_pvname != self._read_pvname and
-                self.setpoint_pvname == pvname):
-            self._run_subs(sub_type=self.SUB_SETPOINT_META,
-                           timestamp=self._metadata['setpoint_timestamp'],
-                           status=self._metadata['setpoint_status'],
-                           severity=self._metadata['setpoint_severity'],
-                           precision=self._metadata['setpoint_precision'],
-                           lower_ctrl_limit=self._metadata['lower_ctrl_limit'],
-                           upper_ctrl_limit=self._metadata['upper_ctrl_limit'],
-                           units=self._metadata['units'],
-                           )
+        if all((from_monitor,
+                self.setpoint_pvname != self.pvname,
+                self.setpoint_pvname == pvname)):
+            # Setpoint has its own metadata
+            self._metadata_thread_ctx.run(
+                self._run_subs,
+                sub_type=self.SUB_SETPOINT_META,
+                timestamp=self._metadata['setpoint_timestamp'],
+                status=self._metadata['setpoint_status'],
+                severity=self._metadata['setpoint_severity'],
+                precision=self._metadata['setpoint_precision'],
+                lower_ctrl_limit=self._metadata['lower_ctrl_limit'],
+                upper_ctrl_limit=self._metadata['upper_ctrl_limit'],
+                units=self._metadata['units'],
+            )
         return metadata
 
-    def _write_changed(self, value=None, timestamp=None, **kwargs):
+    def _write_changed(self, value=None, timestamp=None, from_monitor=True,
```

Comment:
the default of `from_monitor` is not given in `_metadata_changed` and True for `_read_changed`, should these be standardized?  (I would lean to no default in all 3 places).

Reply:
Your comments made me rethink my previous attempt, nuking `write_changed`/`read_changed` `from_monitor=False`.

That also means there's more to review - sorry...